packages/fxa-auth-server/lib/routes/subscriptions/paypal-notifications.ts

```diff
@@ -4,11 +4,68 @@
 import { ServerRoute } from '@hapi/hapi';
 
 import { ConfigType } from '../../../config';
+import { IpnMerchPmtType, isIpnMerchPmt } from '../../payments/paypal-client';
 import { StripeHelper } from '../../payments/stripe';
-import { AuthLogger } from '../../types';
+import { reportSentryError } from '../../sentry';
+import { AuthLogger, AuthRequest } from '../../types';
 import { PayPalHandler } from './paypal';
 
-export class PayPalNotificationHandler extends PayPalHandler {}
+export class PayPalNotificationHandler extends PayPalHandler {
+  private async handleMerchPayment(message: IpnMerchPmtType) {}
+
+  private async handleMpCancel(message: IpnMerchPmtType) {}
+
+  /**
+   * Verify and dispatch IPN events from PayPal
+   *
+   * Run as a fire-and-forget so the execution is wrapped to capture
+   * any errors with the original request context.
+   *
+   * @param request
+   */
+  private async verifyAndDispatchEvent(request: AuthRequest) {
+    try {
+      const verified = await this.paypalHelper.verifyIpnMessage(
+        request.payload as string
+      );
+      if (!verified) {
+        throw new Error('Invalid payload on PayPal IPN Handler.');
+      }
+      const payload = this.paypalHelper.extractIpnMessage(
+        request.payload as string
+      );
+      if (isIpnMerchPmt(payload)) {
+        this.log.debug('Handling Ipn message', { payload });
+        if (payload.txn_type === 'merch_pmt') {
+          return this.handleMerchPayment(payload);
+        } else {
+          return this.handleMpCancel(payload);
+        }
+      }
+      this.log.debug('Unhandled Ipn message', { payload });
+    } catch (err) {
+      reportSentryError(err, request);
+      this.log.error('verifyAndDispatchEvent', {
+        payload: request.payload,
+        err,
+      });
+    }
+    return false;
```

Comment:
Since the return value isn't used, I don't think we need to explicitly return false here.

Reply:
That's true, though it made the tests a little more elegant to verify the function made it to the desired end vs. just 'returning' an undefined implicitly.

Reply:
To me, I interpret that as the function communicating some sort of status with the returned false.  (Obviously if I look at the call site of the function I would know that's not the case.)  YMMV with how others read the intention of that return.